File: PRACTICAS/practica2/jugador.rb
# To change this license header, choose License Headers in Project Properties.
# To change this template file, choose Tools | Templates
# and open the template in the editor.

module Civitas

  class Jugador
    
    @@casas_max = 4
    @@casas_por_hotel = 4
    @@hoteles_max = 4
    @@paso_por_salida = 100
    @@precio_libertad = 200
    @@saldo_inicial = 7500
    
    def initialize(encarcelado = nil, nombre = nil, num_casilla_actual = 0, puede_comprar = false, saldo = @@saldo_inicial, propiedades = [], salvoconducto = nil )
          
        @encarcelado = encarcelado
        @nombre = nombre
        @num_casilla_actual = num_casilla_actual
        @puede_comprar = puede_comprar
        @saldo = saldo
        @propiedades = propiedades
        @salvoconducto = salvoconducto   
        
    end
   
    attr_reader :encarcelado, :num_casilla_actual, :puede_comprar, :salvoconducto, :casas_por_hotel
    
    protected
    attr_reader :nombre, :propiedades, :saldo
    
    private
    
    attr_reader :casas_max, :hoteles_max, :precio_libertad, :paso_por_salida
    
    protected
    
    def self.jugador(jugador)    
      new(jugador.encarcelado, jugador.nombre, jugador.num_casilla_actual, jugador.puede_comprar, jugador.saldo, jugador.propiedades, jugador.salvoconducto)   
    end
    
    
      
    def debe_ser_encarcelado
        
      if !@encarcelado && @salvoconducto == nil         
        return true         
      else
        perder_salvoconducto
        Diario.instance.ocurre_evento("El jugador #{@nombre} se libra de la carcel.")
        return false          
      end
      
    end
    
    public
    
    def encarcelar(num_casilla_carcel)
      if debe_ser_encarcelado        
        mover_a_casilla(num_casilla_carcel)
        @encarcelado = true
        Diario.instance.ocurre_evento("El jugador #{@nombre} ha sido trasladado a la carcel.")
        return @encarcelado     
      end
    end
    
    def obtener_salvoconducto(sorpresa)
      
      if !encarcelado
        @salvoconducto = sorpresa
        return true
      else
        return false
      end
   
    end
    
    private
    
    def perder_salvoconducto
      
      @salvoconducto.usada
      @salvoconducto = nil
      
    end
    
    public
    
    def tiene_salvoconducto
      
      return @salvoconducto != nil
      
    end
    
    def puede_comprar_casilla
      
      if @encarcelado
        @puede_comprar = false
      else
        @puede_comprar = true
      end
      
      return @puede_comprar
      
    end
    
    def paga(cantidad)
      
      return modificar_saldo(cantidad*(-1))
      
    end
    
    def paga_impuesto(cantidad)
      if @encarcelado
        return false
      else
        return paga(cantidad)
      end
    end
    
    def paga_alquiler(cantidad)
      if @encarcelado
        return false
      else
        return paga(cantidad)
      end
    end
    
    def recibe(cantidad)
      if @encarcelado
        return false
      else
        return modificar_saldo(cantidad)
      end
    end
      
    def modificar_saldo(cantidad)
      @saldo += cantidad
      Diario.instance.ocurreEvento("El saldo del jugador #{@nombre} ha sido modificado a #{@saldo}")
      return true
    end
    
    def mover_a_casilla(num_casilla)
      
      if !@encarcelado
        @num_casilla_actual = num_casilla
        @puedo_comprar = false
        Diario.instance.ocurreEvento("El jugador #{@nombre} ha sido movido a la casilla #{@num_casilla_actual}");
        return true
      else
        return false
      end    
      
    end
    
    private
    
    def puedo_gastar(precio)     
      return @saldo >= precio
    end
    
    public
    
    def vender(ip)
      
      if !@encarcelado && existe_la_propiedad(ip)
        vender = propiedades[ip].vender(self)
        
        if vender
          propiedades.delete(ip)
          Diario.instance.ocurre_evento("El jugador #{@nombre} ha vendido la propiedad #{@propiedades[ip].nombre}")
          return true
        end
      end
      
      return false
      
    end
    
    def tiene_algo_que_gestionar
      if @propiedades != null
        return true
      else
        return false
      end
    end
    
    def puede_salir_carcel_pagando
      
      if @saldo >= @precio_libertad
        return true
     
      else
        return false
      
      end
    end
    
    def salir_carcel_pagando
      if @encarcelado && puede_salir_carcel_pagando
        paga(@precio_libertad)
        @encarcelado = false
        Diario.instance.ocurreEvento("El jugador #{@nombre} sale de la carcel pagando");
        return true
      else
        return false
      end
    end
    
    def salir_carcel_tirando
      if Dado.instance.salgo_de_la_carcel
        @encarcelado = false
        Diario.instance.ocurre_evento("El jugador #{@nombre} sale de la carcel")
        return true
      else
        return false
      end
    end
    
    def pasa_por_salida
      @modifica = modificar_saldo(paso_por_salida)
      Diario.instance.ocurre_evento("El jugador #{@nombre} pasa por salida y cobra 100")
      return @modifica
    end
    
    def cantidad_casas_hoteles
      
      suma = 0
      
      for i in (0..@propiedades.size - 1)
        suma += propiedades[i].cantidad_casas_hoteles
      end
      
      return suma
      
    end
    
    def en_bancarrota
      return @saldo < 0
    end
    
    def existe_la_propiedad(ip)
      return ip >= 0 && ip < @propiedades.size
    end
    
    def puedo_edificar_casa(propiedad)
      return true
    end
    
    def puedo_edificar_hotel(propiedad)
      return true
    end
    
    /*
      def cancelar_hipoteca(ip)
      
      end

      def comprar(titulo)

      end

      def construir_casa(ip)

      end

      def construir_hotel(ip)

      end

      def hipotecar(ip)

      end
    
    */
      
    def to_string(ip)       
      return "Jugador de nombre #{@nombre} situado en la casilla #{@num_casilla_actual} con saldo #{@saldo}";
    end
    
    
  end
    
    
end

File: PRACTICAS/CAMBIAR_RUBY/COMPROBACIONES JAVA/src/civitas_p1/Jugador.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package civitas_p1;

import static java.lang.Float.compare;
import java.util.ArrayList;

/**
 *
 * @author angel
 */

    
public class Jugador implements Comparable<Jugador>{
    
    protected static final int CASASMAX = 4;
    protected static final int CASASPORHOTEL = 4;
    protected boolean encarcelado;
    protected static final int HOTELESMAX = 4;
    private String nombre;
    private int numCasillaActual;
    protected static final float PASOPORSALIDA = 100;
    protected static final float PRECIOLIBERTAD = 200;
    private boolean puedeComprar;
    private float saldo;
    private static final float SALDOINICIAL = 7500;
    private ArrayList<TituloPropiedad> propiedades;
    private Sorpresa salvoconducto;
    
    @Override
    public int compareTo(Jugador otro) {
        return compare(saldo, otro.saldo);
    }
    
    Jugador(String nombre){
        
        encarcelado = false;
        this.nombre = nombre;
        numCasillaActual = 0;
        puedeComprar = true;
        saldo = SALDOINICIAL;
        propiedades = new ArrayList<TituloPropiedad>();
        salvoconducto = null;
        
    }
    
    protected Jugador (Jugador jugador){
        encarcelado = jugador.encarcelado;
        nombre = jugador.nombre;
        numCasillaActual = jugador.numCasillaActual;
        puedeComprar = jugador.puedeComprar;
        saldo = jugador.saldo;
        salvoconducto = jugador.salvoconducto;
        
        propiedades = new ArrayList<>();
        
        if (!jugador.propiedades.isEmpty())
            for (int i = 0; i < jugador.propiedades.size(); i++)
                propiedades.set(i, jugador.propiedades.get(i));
        else
            propiedades = null;
        
    }
    
    protected boolean debeSerEncarcelado(){
        
        if (!encarcelado){
            if (salvoconducto == null)
                return true;
            else{
                perderSalvoconducto();
                Diario.instance.ocurreEvento("El jugador " + nombre + " se libra de la carcel.");
                return false;
                
            }
                
        }
        else
            return false;
        
    }
    
    boolean encarcelar(int numCasillaCarcel){
        if (debeSerEncarcelado()){
            moverACasilla(numCasillaCarcel);
            encarcelado = true;
            Diario.instance.ocurreEvento("El jugador " + nombre + " ha sido trasladado a la carcel.");
        }
        
        return encarcelado;
    }
    
    boolean obtenerSalvoconducto(Sorpresa sorpresa){
        
        if (!encarcelado){
            salvoconducto = sorpresa;
            return true;
        }
        else
            return false;
    }
    
    private void perderSalvoconducto(){
        salvoconducto.usada();
        salvoconducto = null;
    }
    
    boolean tieneSalvoconducto(){
        return salvoconducto != null;
    }
    
    boolean puedeComprarCasilla(){
        puedeComprar = !encarcelado;
        System.out.println("\nENTRA EN EL PUEDE COMPRAR, LO PONE A " + puedeComprar);
        return puedeComprar;
    }
    
    boolean paga(float cantidad){
        return modificarSaldo(cantidad*(-1));
    }
    
    boolean pagaImpuesto(float cantidad){
        if (encarcelado)
            return false;
        else
            return paga(cantidad);
    }
    
    boolean pagaAlquiler(float cantidad){
        if (encarcelado)
            return false;
        else
            return paga(cantidad);
    }
    
    boolean recibe(float cantidad){
        if (encarcelado)
            return false;
        else
            return modificarSaldo(cantidad);
    }
    
    boolean modificarSaldo(float cantidad){
        saldo += cantidad;
        Diario.instance.ocurreEvento("El saldo del jugador " + nombre + " ha sido modificado a " + saldo);
        return true;
    }
    
    boolean moverACasilla(int numCasilla){
        if (!encarcelado){
            numCasillaActual = numCasilla;
            Diario.instance.ocurreEvento("El jugador " + nombre + " ha sido movido a la casilla " + numCasilla);
           return true;
        }
        else
            return false;
    }
    
    private boolean puedoGastar(float precio){
        if (!encarcelado)
            return saldo >= precio;
        else
            return false;
    }
    
    boolean vender(int ip){
        
        if (!encarcelado){
            if(existeLaPropiedad(ip)){
                boolean vender = propiedades.get(ip).vender(this);
                if(vender){
                    propiedades.remove(ip);
                    Diario.instance.ocurreEvento("El jugador " + nombre + " ha vendido la propiedad " + propiedades.get(ip).getNombre());
                    return true;
                }
                else
                    return false;
            }
            else
                return false;
        }
        else
            return false;
    }
    
    boolean tieneAlgoQueGestionar(){
        return propiedades != null;
    }
    
    boolean puedeSalirCarcelPagando(){
        return saldo >= PRECIOLIBERTAD;
    }
    
    boolean salirCarcelPagando(){
        if (encarcelado && puedeSalirCarcelPagando()){
            paga(PRECIOLIBERTAD);
            encarcelado = false;
            Diario.instance.ocurreEvento("El jugador " + nombre + " sale de la carcel pagando");
            return true;
        }
        else
            return false;
        
    }
    
    boolean salirCarcelTirando(){
        if(Dado.instance.salgoDeLaCarcel()){
            encarcelado = false;
            Diario.instance.ocurreEvento("El jugador " + nombre + " sale de la carcel.");
            return true;
        }   
        else
            return false;
    }
    
    boolean pasaPorSalida(){
        boolean modifica;
        modifica = modificarSaldo(PASOPORSALIDA);
        Diario.instance.ocurreEvento("El jugador " + nombre + " pasa por salida y cobra 100.");
        return modifica;
    }
    
    int cantidadCasasHoteles(){
        
        int suma = 0;
        
        for (int i = 0; i < propiedades.size(); i++)
            suma += propiedades.get(i).cantidadCasasHoteles();
        
        return suma;
    }
    
    boolean enBancarrota(){
        return saldo < 0;
    }
    
    private boolean existeLaPropiedad(int ip){
        return ip >= 0 && ip < propiedades.size();
    }
    
    private int getCasasMax(){
        return CASASMAX;
    }
    
    int getCasasPorHotel(){
        return CASASPORHOTEL;
    }
    
    private int getHotelesMax(){
        return HOTELESMAX;
    }
    
    protected String getNombre(){
        return nombre;
    }
    
    int getNumCasillaActual(){
        return numCasillaActual;
    }
    
    private float getPrecioLibertad(){
        return PRECIOLIBERTAD;
    }
    
    private float getPremioPasoPorSalida(){
        return PASOPORSALIDA;
    }
    
    public ArrayList<TituloPropiedad> getPropiedades(){
        return propiedades;
    }
    
    boolean getPuedeComprar(){
        return puedeComprar;
    }
    
    protected float getSaldo(){
        return saldo;
    }
    
    public boolean isEncarcelado(){
        return encarcelado;
    }
    
    private boolean puedoEdificarCasa(TituloPropiedad propiedad){
        
        float precio = propiedad.getPrecioEdificar();
        
        return propiedad.esEsteElPropietario(this) && propiedad.getNumCasas() < CASASPORHOTEL && this.puedoGastar(precio);
        
    }
    
    private boolean puedoEdificarHotel(TituloPropiedad propiedad){
        
        float precio = propiedad.getPrecioEdificar();
        
        return propiedad.getNumCasas() >=  CASASPORHOTEL && propiedad.getNumHoteles() < CASASPORHOTEL && this.puedoGastar(precio);
    }
    
    boolean cancelarHipoteca(int ip){
    
        boolean result=false;
        
        if(encarcelado){
            
            return result;
            
        }
        if(existeLaPropiedad(ip)){
            
            TituloPropiedad propiedad = propiedades.get(ip);  
            float cantidad = propiedad.getImporteCancelarHipoteca();
            boolean puedoGastar=puedoGastar(cantidad);
            if(puedoGastar){
                
                result=propiedad.cancelarHipoteca(this);
            
                
                if(result){
                    
                Diario.getInstance().ocurreEvento("El jugador "+nombre+ " cancela la hipoteca de la propiedad " +propiedades.get(ip).getNombre());
                    
                }
                
                
            }
            
        }
        
        
        return result;
    
    }
    
    
       boolean comprar(TituloPropiedad titulo){
           
           boolean result = false;

        if (encarcelado)
            return result;


        if (puedeComprar){

            float precio = titulo.getPrecioCompra();

            if (puedoGastar(precio)){

                result = titulo.comprar(this);

                if (result){
                    propiedades.add(titulo);
                    Diario.instance.ocurreEvento("El jugador "+nombre+ "compra la propiedad" + titulo.toString());
                }
                
                puedeComprar = false ;
                
             }

       }

        return result;
           
           
        }
    
     
    
    Boolean construirHotel (int ip){
    
    boolean result = false;
        boolean puedoEdificarHotel;
        float precio;
        int casasPorHotel;
        
        if(encarcelado)
            return result;
        
        if(existeLaPropiedad(ip)){
            
            
            TituloPropiedad propiedad;
            propiedad = propiedades.get(ip);
            
            
            puedoEdificarHotel=puedoEdificarHotel(propiedad);
            
            if(puedoEdificarHotel){
                
                result = propiedad.construirHotel(this);
                
                
                casasPorHotel=getCasasPorHotel();
                
                propiedad.derruirCasas(casasPorHotel, this);
                
            }
            
            Diario.getInstance().ocurreEvento("El jugador"+nombre+" costruye hotel en la propiedad"+ip);
        }
        
        return result;
    
    }
    
    
    Boolean construirCasa(int ip){
        
        boolean result =false, puedoEdificarCasa =false;
        
        if(encarcelado){
            return result;
        }else{
            
            boolean existe = existeLaPropiedad(ip);
        
            if(existe){
                
                TituloPropiedad propiedad=propiedades.get(ip);
                puedoEdificarCasa=puedoEdificarCasa(propiedad);
                float precio=propiedad.getPrecioEdificar();
                if(puedoGastar(precio) && propiedad.getNumCasas()<getCasasMax()){
                    
                    puedoEdificarCasa=true;
                    
                }
                if(puedoEdificarCasa){
                    
                    result=propiedad.construirCasa(this);
                    
                    if(result){
                        
                        Diario.getInstance().ocurreEvento("El jugador " +nombre+ " construye casa en la propiedad"+ propiedades.get(ip));
                        
                    }
                }
            }
        }
        
        return result;
    }
    
    protected boolean hipotecar(int ip){
        
        boolean result=false;
        if(encarcelado){
            
            return result;
            
        }
        
        if(existeLaPropiedad(ip)){
            
            TituloPropiedad propiedad = propiedades.get(ip); 
            result=propiedad.hipotecar(this);
                     
        }
        if(result){
            
            Diario.getInstance().ocurreEvento("El jugador " +nombre+ " hipoteca la propiedad "+ propiedades.get(ip).getNombre());
        }
        
        return result;
    }
    
    
    @Override
    public String toString(){
        
        return "Jugador de nombre " + nombre + " situado en la casilla " + numCasillaActual + " con saldo " + saldo;
    }
    
    
}
